### ISSUE TITLE
Improve interaction between optional auto-feature subprojects with auto-feature options

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#5688]
Let's say you have an optional subproject:

```meson
# meson.build
subproject('foo', required: get_option('foo-opt'))
...
# meson_options.txt
option('foo-opt', type: 'feature', value : 'auto')
```

Now let's say subproject `foo` has an optional dependency:

```meson
# meson.build
bar_dep = dependency('bar', required: get_option('bar-opt'))
...
# meson_options.txt
option('bar-opt', type: 'feature', value: 'auto')
```

Now you call meson like this:

```
$ meson _build -Dfoo:bar-opt=enabled
```

Since you explicitly enabled an option, you would expect meson to error out if that option could not be enabled. However, since the subproject is optional, meson will print an error, ***skip the subproject*** entirely, and continue.

I think this is quite unexpected behaviour and is easy to miss. I would expect that what the user wants in such a situation is `'I want this option enabled, and if it can't be, I want meson to error out'`, like with feature options everywhere else.

I tried to concoct a patch that implements this, but this is a somewhat tricky edge-case, hence I'm documenting all this in an issue:

We don't maintain 'layers' of option values, so we have no way to differentiate between 'option is enabled on the command-line / in configuration' vs 'option default is enabled'. We only want this behaviour in the former case, not the latter case.

Worse, the mapping between subproject and its feature option is defined inside the build file, so we can't do this at option-parsing time. Must be done inside the interpreter; perhaps while handling exceptions raised by a subproject and special-casing `DependencyException` from a feature option that was overriden to be 'enabled'.

**Comment by xclaesse** [comment]
I agree with this issue, it hit me a few times with gst-build. But indeed the implementation could be tricky. Note that you can know which options have been set from command line using `read_cmd_line_file()`.

When we raise DependencyException we could set a flag on it saying if it's because of a enabled feature, `dependency_impl()` knows which feature has been passed.

**Comment by nirbheek** [comment]
> Note that you can know which options have been set from command line using `read_cmd_line_file()`

Does that include options set using `meson configure` on the builddir afterwards?

**Comment by nirbheek** [comment]
> When we raise DependencyException we could set a flag on it saying if it's because of a enabled feature, `dependency_impl()` knows which feature has been passed.

I tried exactly this when I was trying to implement this, and it can be done, but currently we 'squash' that information into a boolean in the interpreter, and `dependencyes/base.py:find_external_dependency()` only sees `required: {true,false}`. That can be changed, but I wasn't sure if this whole idea was a good way to do things.

**Comment by xclaesse** [comment]
> > Note that you can know which options have been set from command line using `read_cmd_line_file()`
> 
> Does that include options set using `meson configure` on the builddir afterwards?

It contains all option set by command line at initial setup and those set with meson configure afterward. Note it won't contain options set in current invokation until the very end. So what you can do is read that file at the beggining of mesonmain, update with current options, and store that dict on environment. That way we have the dict of options explicitely set by user.
